### PR TITLE
SERVICES-223, changed consul class to take in multiple folder paths

### PR DESCRIPTION
**Pull request by connietruong**
@nmonterroso 
The update changes the data structure DefaultConsulKeyValueConfig so it can now take various folder paths to hold key values. There are three paths based on priority and commonalities with other possible key values. The original constructors still remain.

The paths include the original config/app/env path, but additionally includes a config/base/env and config/app/base for easier sorting and searching.

The new implementation passed the test written for the previous implementation, and another test that tests the new constructor made to hold multiple folders.

### CODE REVIEW

**Comment by nmonterroso** [comment]
+ @Wikia/services-team

File: lib/consul-config/src/main/java/com/wikia/gradle/DefaultConsulKeyValueConfig.java [hunk 1]
```diff
@@ -19,16 +21,35 @@
...
+  }
+
+  public DefaultConsulKeyValueConfig(List<String> configFolders) {
+    this.configFolders = configFolders;
+  }
+
+  protected static List<String> getFolderList(List<ConfigFolderLocation> folders) {
+    List<String> tempFolder = new ArrayList<>();
+    String folder;
+    for (ConfigFolderLocation v : folders) {
```
Comment (harnash) [inline lib/consul-config/src/main/java/com/wikia/gradle/DefaultConsulKeyValueConfig.java]: I would choose better name ```v -> config_folder``` or something similar ```v``` is not descriptive enough.

File: lib/consul-config/src/main/java/com/wikia/gradle/DefaultConsulKeyValueConfig.java [hunk 2]
```diff
@@ -19,16 +21,35 @@
...
+      folder = getDefaultConfigFolder(v.app, v.env);
+      LOG.info("Reading configuration from consul path: {}", folder);
```
Comment (harnash) [inline lib/consul-config/src/main/java/com/wikia/gradle/DefaultConsulKeyValueConfig.java]: Might be worth printing out the whole list in one run? Perhaps using construction ```(Arrays.toString(tempFolder.toArray()))```

File: lib/consul-config/src/main/java/com/wikia/gradle/DefaultConsulKeyValueConfig.java [hunk 3]
```diff
@@ -19,16 +21,35 @@
...
+    this(getFolderList(folders));
+    this.consul = consul;
+  }
+
+  public DefaultConsulKeyValueConfig(List<String> configFolders) {
+    this.configFolders = configFolders;
+  }
+
+  protected static List<String> getFolderList(List<ConfigFolderLocation> folders) {
+    List<String> tempFolder = new ArrayList<>();
```
Comment (harnash) [inline lib/consul-config/src/main/java/com/wikia/gradle/DefaultConsulKeyValueConfig.java]: ```tempFolder``` -> ```tempFolders``` or perhaps ```configFolders```?

File: lib/consul-config/src/main/java/com/wikia/gradle/DefaultConsulKeyValueConfig.java [hunk 4]
```diff
@@ -41,21 +62,23 @@ protected static String getDefaultConfigFolder(String app, String env) {
...
-            config.put(key, value);
-          } catch (Exception e) {
-            LOG.error("Could not fetch consul value for key \"{}\" from folder {}", key,
-                      configFolder, e);
-          }
-        });
+    for (String configLocation : configFolders) {
+      getRawConfigValues(configLocation).map(List::stream).orElse(Stream.<Value>empty())
+              .filter(v -> getConfigKey(v.getKey(), configLocation) != null) // returns null for folders
+              .filter(v -> !v.getKey().equals(configLocation))
```
Comment (harnash) [inline lib/consul-config/src/main/java/com/wikia/gradle/DefaultConsulKeyValueConfig.java]: Two above filters could be combined into one.

**Comment by harnash** [comment]
Except some minor comments - LGTM but please wait for someone else to look at this code.

File: lib/consul-config/src/main/java/com/wikia/gradle/DefaultConsulKeyValueConfig.java [hunk 5]
```diff
@@ -19,16 +21,35 @@
   private final static Logger LOG = LoggerFactory.getLogger(DefaultConsulKeyValueConfig.class);
   protected Consul consul;
   protected HashMap<String, String> config;
-  protected String configFolder;
+  protected List<String> configFolders;
```
Comment (nmonterroso) [inline lib/consul-config/src/main/java/com/wikia/gradle/DefaultConsulKeyValueConfig.java]: can we add a comment here saying that the order of the folders matters, and how?

**Pull request merged**